File: app/src/main/java/com/happs/ximand/ringcontrol/view/fragment/BaseFragmentWithRecyclerView.java
package com.happs.ximand.ringcontrol.view.fragment;

import androidx.annotation.NonNull;
import androidx.databinding.ViewDataBinding;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.RecyclerView.ViewHolder;

import com.happs.ximand.ringcontrol.view.BaseFragment;
import com.happs.ximand.ringcontrol.view.adapter.BaseRecyclerViewAdapter;
import com.happs.ximand.ringcontrol.viewmodel.fragment.BaseViewModel;

import java.util.List;

public abstract class BaseFragmentWithRecyclerView<VM extends BaseViewModel,
        D extends ViewDataBinding, T, A extends BaseRecyclerViewAdapter<T, ? extends ViewHolder>>
        extends BaseFragment<VM, D> {

    private RecyclerView recyclerView;

    public BaseFragmentWithRecyclerView(int layoutId, int menuResId) {
        super(layoutId, menuResId);
    }

    @Override
    protected void onViewDataBindingCreated(@NonNull D binding) {
        this.recyclerView = getRecyclerViewFromBinding(binding);
        this.recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
    }

    protected abstract RecyclerView getRecyclerViewFromBinding(D binding);

    protected void initAdapter(List<T> items) {
        if (items == null) {
            throw new IllegalStateException();
        }
        A adapter = getAdapter();
        A newAdapter = createNewAdapter(items);
        onPreAttachRecyclerViewAdapter(newAdapter);
        recyclerView.setAdapter(newAdapter);
        if (adapter != null) {
            recyclerView.invalidate();
        }
    }

    protected abstract A createNewAdapter(List<T> items);

    protected void onPreAttachRecyclerViewAdapter(A adapter) {

    }

    @SuppressWarnings("unchecked")
    protected A getAdapter() {
        return (A) recyclerView.getAdapter();
    }

}

File: app/src/main/java/com/happs/ximand/ringcontrol/view/fragment/BaseFragmentWithRecyclerView.kt
package com.happs.ximand.ringcontrol.view.fragment

import androidx.databinding.ViewDataBinding
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.happs.ximand.ringcontrol.view.adapter.BaseRecyclerViewAdapter
import com.happs.ximand.ringcontrol.view.adapter.BaseViewHolder
import com.happs.ximand.ringcontrol.viewmodel.fragment.BaseViewModel

abstract class BaseFragmentWithRecyclerView<VM : BaseViewModel, B : ViewDataBinding, T, RB : ViewDataBinding,
        A : BaseRecyclerViewAdapter<T, out BaseViewHolder<T, RB>, RB>>(layoutId: Int, menuResId: Int)
    : BaseFragment<VM, B>(layoutId, menuResId) {

    private var recyclerView: RecyclerView? = null

    @Suppress("UNCHECKED_CAST")
    protected val adapter: A?
        get() = recyclerView!!.adapter as A?

    override fun onViewDataBindingCreated(binding: B) {
        recyclerView = getRecyclerViewFromBinding(binding)
        recyclerView!!.layoutManager = LinearLayoutManager(context)
    }

    protected abstract fun getRecyclerViewFromBinding(binding: B): RecyclerView?

    protected fun initAdapter(items: MutableList<T>?) {
        checkNotNull(items)
        val adapter = adapter
        val newAdapter = createNewAdapter(items)
        onPreAttachRecyclerViewAdapter(newAdapter)
        recyclerView!!.adapter = newAdapter
        if (adapter != null) {
            recyclerView!!.invalidate()
        }
    }

    protected abstract fun createNewAdapter(items: MutableList<T>): A

    protected open fun onPreAttachRecyclerViewAdapter(adapter: A) {}
}